onnxruntime/core/session/provider_bridge_ort.cc

```diff
@@ -0,0 +1,563 @@
+    if (attributes_.empty()) {
+      for (auto& v : p_->GetAttributes())
+        attributes_[v.first] = onnxruntime::make_unique<Provider_AttributeProto_Impl>(v.second);
+    }
+    return attributes_;
+  }
+
+  size_t GetInputEdgesCount() const noexcept override {
+    return p_->GetInputEdgesCount();
+  }
+  size_t GetOutputEdgesCount() const noexcept override { return p_->GetOutputEdgesCount(); }
+
+  std::unique_ptr<Provider_NodeIterator> InputNodesBegin_internal() const noexcept override;
+  std::unique_ptr<Provider_NodeIterator> InputNodesEnd_internal() const noexcept override;
+
+  const Node* p_;
+  mutable std::vector<Provider_NodeArg*> input_defs_;
+  mutable std::vector<Provider_NodeArg*> output_defs_;
+  mutable Provider_NodeAttributes attributes_;
+};
+
+struct Provider_NodeIterator_Impl : Provider_Node::Provider_NodeIterator {
+  Provider_NodeIterator_Impl(Node::NodeConstIterator&& v) : v_{std::move(v)} {}
+
+  bool operator!=(const Provider_NodeIterator& p) const override { return v_ != static_cast<const Provider_NodeIterator_Impl*>(&p)->v_; }
+
+  void operator++() override { return v_.operator++(); }
+  const Provider_Node& operator*() override {
+    node_ = Provider_Node_Impl(&*v_);
+    return node_;
+  }
+
+  Node::NodeConstIterator v_;
+  Provider_Node_Impl node_{nullptr};
+};
+
+std::unique_ptr<Provider_Node::Provider_NodeIterator> Provider_Node_Impl::InputNodesBegin_internal() const noexcept {
+  return onnxruntime::make_unique<Provider_NodeIterator_Impl>(p_->InputNodesBegin());
+}
+
+std::unique_ptr<Provider_Node::Provider_NodeIterator> Provider_Node_Impl::InputNodesEnd_internal() const noexcept {
+  return onnxruntime::make_unique<Provider_NodeIterator_Impl>(p_->InputNodesEnd());
+}
+
+struct Provider_IndexedSubGraph_Impl : Provider_IndexedSubGraph {
+  Provider_IndexedSubGraph_Impl() = default;
+  Provider_IndexedSubGraph_Impl(std::unique_ptr<IndexedSubGraph> p) : p_{std::move(p)} {}
+
+  void SetMetaDef(std::unique_ptr<MetaDef>& def_) override {
+    auto real = onnxruntime::make_unique<IndexedSubGraph::MetaDef>();
+
+    real->name = std::move(def_->name);
+    real->domain = std::move(def_->domain);
+    real->since_version = def_->since_version;
+    real->status = def_->status;
+    real->inputs = std::move(def_->inputs);
+    real->outputs = std::move(def_->outputs);
+
+    for (auto& v : def_->attributes)
+      real->attributes.emplace(v.first, static_cast<Provider_AttributeProto_Impl*>(v.second.p_.get())->v_);
+
+    real->doc_string = std::move(def_->doc_string);
+
+    p_->SetMetaDef(real);
+  }
+
+  std::vector<onnxruntime::NodeIndex>& Nodes() override { return p_->nodes; }
+
+  std::unique_ptr<IndexedSubGraph> p_{onnxruntime::make_unique<IndexedSubGraph>()};
+};
+
+struct Provider_GraphViewer_Impl : Provider_GraphViewer {
+  Provider_GraphViewer_Impl(const GraphViewer& v) : v_(v) {
+    for (int i = 0; i < v_.MaxNodeIndex(); i++)
+      provider_nodes_.emplace_back(v_.GetNode(i));
+  }
```

Comment:
nit: prefer push_back. [https://abseil.io/tips/112](https://abseil.io/tips/112)

Reply:
Hmm, but in this case I want to call the constructor as the vector is of type Provider_Node_Impl, and I'm passing a Node* to its constructor explicitly this way. It's not a vector of Node*.

I can try changing it, but does that justify it?